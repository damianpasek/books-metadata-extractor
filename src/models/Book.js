import Sequelize from 'sequelize'
import db from '../config/db'

const Book = db.define('Book', {
  id: {
    type: Sequelize.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  external_id: {
    type: Sequelize.INTEGER,
  },
  title: {
    type: Sequelize.STRING(1024),
  },
  author: {
    type: Sequelize.STRING(256),
  },
  publisher: {
    type: Sequelize.STRING(64),
    allowNull: false,
  },
  publication_date: {
    type: Sequelize.DATE,
  },
  language: {
    type: Sequelize.STRING(2),
  },
  subjects: {
    type: Sequelize.JSON,
  },
  license: {
    type: Sequelize.STRING(128),
  },
}, {
  tableName: 'books',
  underscored: true,
  timestamps: false,
})

module.exports = Book
