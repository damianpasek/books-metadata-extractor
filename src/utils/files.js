import path from 'path'
import glob from 'glob'
import fs from 'fs'
import { Parser } from 'xml2js'

const parser = new Parser()

export const getDataDirectoryPath = () => {
  const dirPath = process.env.FILES_PATH || path.join(__dirname, '../..', 'data')
  return `${dirPath}/**/*.rdf`
}

export const getFilesPaths = path => new Promise((resolve, reject) => {
  glob(path, {}, (error, files) => {
    if (error) reject(error)

    const limit = parseInt(process.env.FILES_LIMIT) || files.length
    resolve(files.slice(0, limit))
  })
})

export const getBookDataFromFile = filePath => new Promise((resolve, reject) => {
  fs.readFile(filePath, { encoding: 'utf-8' }, (error, data) => {
    if (error) return reject(error)

    parser.parseString(data, (error, parsed) => {
      if (error) return reject(error)
      resolve(parsed)
    })
  })
})
