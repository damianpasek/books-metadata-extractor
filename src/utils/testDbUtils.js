import db from '../config/db'

export const syncDb = async () => {
  await db.drop()
  await db.sync({ force: true })
}

export const clearDb = () => db.drop()
