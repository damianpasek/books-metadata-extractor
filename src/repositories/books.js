import { chunk } from 'lodash'
import Book from '../models/Book'

const BATCH_SIZE = process.env.BATCH_SIZE || 50

export const saveBooksInDatabase = async (books = []) => {
  const batches = chunk(books, BATCH_SIZE)

  for (const batch of batches) {
    await Book.bulkCreate(batch)
  }
}
