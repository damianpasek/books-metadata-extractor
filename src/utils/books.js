import { get } from 'lodash'

import { getBookDataFromFile } from './files'

const getProperty = (data, path = []) => get(data, path, null)

const getIdFromFilePath = path => {
  const parsed = path.split('/').pop().split('.')[0].replace('pg', '')

  return parsed ? parseInt(parsed) : null
}

const getSubjects = book => {
  const subjects = getProperty(book, ['dcterms:subject']) || []

  return subjects
    .map(subject => getProperty(subject, ['rdf:Description', 0, 'rdf:value', 0]))
    .filter(Boolean)
}

const parseBookData = (id, data) => {
  const book = data['rdf:RDF']['pgterms:ebook'][0]

  const title = getProperty(book, ['dcterms:title', 0])
  const author = getProperty(book, ['dcterms:creator', 0, 'pgterms:agent', 0, 'pgterms:name', 0])
  const language = getProperty(book, ['dcterms:language', 0, 'rdf:Description', 0, 'rdf:value', 0, '_'])
  const license = getProperty(book, ['dcterms:rights', 0])
  const publisher = getProperty(book, ['dcterms:publisher', 0])
  const publicationDate = getProperty(book, ['dcterms:issued', 0, '_'])
  const subjects = getSubjects(book)

  return {
    external_id: id,
    title,
    author,
    publisher,
    publication_date: publicationDate,
    language,
    subjects,
    license,
  }
}

export const getBookData = async (filePath) => {
  const id = getIdFromFilePath(filePath)
  const fileContent = await getBookDataFromFile(filePath)

  return parseBookData(id, fileContent)
}
