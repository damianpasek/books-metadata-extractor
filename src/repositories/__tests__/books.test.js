import { saveBooksInDatabase } from '../books'
import { syncDb, clearDb } from '../../utils/testDbUtils'
import Book from '../../models/Book'
import { mockedBooks } from '../__mocks__/books.mock'

describe('Repositories: books', () => {
  describe('saveBooksInDatabase', () => {
    beforeEach(async () => {
      await syncDb()
    })

    afterEach(async () => {
      await clearDb()
    })

    it('should save given list of books in the database', async () => {
      await saveBooksInDatabase(mockedBooks)

      const books = await Book.findAll()
      const book = await Book.findOne({ where: { id: 1 } })

      expect(books.length).toBe(10)
      expect(books).toMatchSnapshot()
      expect(book).toMatchObject({
        id: 1,
        external_id: 10001,
        title: 'Apocolocyntosis',
        author: 'Seneca, Lucius Annaeus',
        publisher: 'Project Gutenberg',
        language: 'en',
        subjects: [
          'PA',
          'Claudius, Emperor of Rome, 10 B.C.-54 A.D. -- Humor',
        ],
        license: 'Public domain in the USA.',
      })
    })

    it('should throw error on database constraint', async () => {
      const mockedBook = [{ title: 'Mocked Title' }]

      await expect(saveBooksInDatabase(mockedBook))
        .rejects
        .toThrowError()
    })
  })
})
