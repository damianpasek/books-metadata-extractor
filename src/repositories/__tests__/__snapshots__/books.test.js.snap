// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories: books saveBooksInDatabase should save given list of books in the database 1`] = `
Array [
  Object {
    "author": "Seneca, Lucius Annaeus",
    "external_id": 10001,
    "id": 1,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 2003-11-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "PA",
      "Claudius, Emperor of Rome, 10 B.C.-54 A.D. -- Humor",
    ],
    "title": "Apocolocyntosis",
  },
  Object {
    "author": "Jefferson, Thomas",
    "external_id": 1,
    "id": 2,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 1971-12-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "JK",
      "United States. Declaration of Independence",
      "United States -- History -- Revolution, 1775-1783 -- Sources",
      "E201",
    ],
    "title": "The Declaration of Independence of the United States of America",
  },
  Object {
    "author": "Dante Alighieri",
    "external_id": 1000,
    "id": 3,
    "language": "it",
    "license": "Public domain in the USA.",
    "publication_date": 1997-08-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "Epic poetry, Italian",
      "Italian poetry -- To 1400",
      "PQ",
    ],
    "title": "La Divina Commedia di Dante: Complete",
  },
  Object {
    "author": null,
    "external_id": 10,
    "id": 4,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 1989-08-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "BS",
      "Bible",
    ],
    "title": "The King James Version of the Bible",
  },
  Object {
    "author": "Shakespeare, William",
    "external_id": 100,
    "id": 5,
    "language": "en",
    "license": "Copyrighted. Read the copyright notice inside this book for details.",
    "publication_date": 1994-01-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "English drama -- Early modern and Elizabethan, 1500-1600",
      "PR",
    ],
    "title": "The Complete Works of William Shakespeare",
  },
  Object {
    "author": "Hodgson, William Hope",
    "external_id": 10002,
    "id": 6,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 2003-11-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "Science fiction",
      "PR",
    ],
    "title": "The House on the Borderland",
  },
  Object {
    "author": "Lindsay, Anna Robertson Brown",
    "external_id": 10004,
    "id": 7,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 2003-11-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "Christianity",
      "BR",
    ],
    "title": "The Warriors",
  },
  Object {
    "author": "Waddington, Mary King",
    "external_id": 10003,
    "id": 8,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 2003-11-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "France -- Social life and customs",
      "DC",
      "France -- History -- Third Republic, 1870-1940",
    ],
    "title": "My First Years as a Frenchwoman, 1876-1879",
  },
  Object {
    "author": "Tucker, George",
    "external_id": 10005,
    "id": 9,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 2003-11-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "Science fiction",
      "PS",
      "Space flight to the moon -- Fiction",
    ],
    "title": "A Voyage to the Moon
With Some Account of the Manners and Customs, Science and Philosophy, of the People of Morosofia, and Other Lunarians",
  },
  Object {
    "author": "Anonymous",
    "external_id": 10000,
    "id": 10,
    "language": "en",
    "license": "Public domain in the USA.",
    "publication_date": 2006-03-01T00:00:00.000Z,
    "publisher": "Project Gutenberg",
    "subjects": Array [
      "Constitutional history -- England -- Sources",
      "KD",
      "Magna Carta",
      "JN",
    ],
    "title": "The Magna Carta",
  },
]
`;
